tensorboard/webapp/metrics/store/metrics_store_internal_utils_test.ts

```diff
@@ -502,4 +504,104 @@ describe('metrics store utils', () => {
       expect(canCreateNewPins(state)).toBe(false);
     });
   });
+
+  describe('updateCardMaps', () => {
```

Comment:
Although I appreciate the modularity of breaking this up into updateCardMaps(), I think at least some of the testing for this use case needs to be done at the reducer level. This new logic is a core but subtle behavior of a very complex reducer.

* Keeping it disjoint like this actually makes it more difficult to reason out what the behavior of the reducer is supposed to be.

* Keeping it disjoint makes the reducer more difficult to successfully refactor without introducing bugs. A key goal should be: Someone should be able to refactor the reducer logic and have strong certainty that these use cases haven't been broken. (I personally would like to hack around with this reducer code to see if we can greatly simplify it and having the tests at the reducer level would make this easier).

Reply:
Just want to clarify: I think its fine to keep updateCardMaps() but I think some of the testing needs to be moved or duplicated in metrics_reducers_test.

Reply:
Thanks for the clarification and suggestion. I was thinking of the duplicated tests too and thought maybe they should not be duplicated. On the other hand, it also does not hurt to add more tests. I will just add tests for the method for future easiness of refactoring.